### ISSUE TITLE
Introduce common undo functionality and prevent accidental cell deletion

### ISSUE THREAD

**Issue by repetitioestmaterstudiorum** [microsoft/vscode#165159]
My request concerns the same as this one (closed issue from 2+ years ago): https://github.com/microsoft/vscode-jupyter/issues/1998

It seems to be a frequently occurring problem until today, and a reason for people not to use VSCode for Jupyter notebooks. See for example: https://stackoverflow.com/questions/61603833/how-to-recover-accidentally-deleted-cells-in-jupyter-notebook

It happens all too often: once a cell is not in focus, and one hits the common key combination to undo the last change (on Mac: Cmd + z), a cell gets deleted. Despite the suggested "solution" in issue 1998 (just hitting the key "z" when not focused on a cell), this does more of the same: it deletes cells!
Once a cell is deleted, undoing via menu or via Cmd + Shift + P (command palette), deletes just more cells. 

So basically, with this extension, undoing in any way means deleting. And deleting can't be undone. 

I use VSCode for programming a lot. Just removing the common undo key combination is not an option for me. 

Also, I think there really ought to be a working solution to undo an accidental deletion, because it seems to be the norm in any application these days to have the ability to undo the last change (deleting a cell is a change as well).

Besides the obvious: it would be nice to undo the last edit, whether a cell is active or not!

Has there been any developments in this regard?

**Comment by repetitioestmaterstudiorum** [comment]
Exactly @roblourens it's about the undo operation (no matter how it's called - "z", "cmd + z", from the command palette, from the menu), and that it doesn't undo the last action but deletes a cell, if the cell is not focused.

BUT: this also happens when there IS focus on a cell, one scrolls up to another part of the notebook, and types. Usually, VSCode jumps back to where the cursor is still active (as do other apps). In jupyter mode, the last active cell is deleted. 
And yes, if one "panicks", then indeed, more cells are added to the "kill list" ;)

_I use Local history, but it does not record any history when working on jupyter notebooks, and I found no config to change that. Also jupyter checkpoints don't get created (I saw another issue about that - I believe that is related to my M1 device though, so let's not entangle issues)._

I think there are the following two problems here:

1. the undo action behaves very abnormally compared to any other modern app, when a cell is not active
2. the same applies if there theoretically IS focus on a cell but the cell is scrolled out of the viewport
3. once undo accidentally deletes a cell, there's no way of restoring it

**Comment by roblourens** [comment]
It is undoing the last notebook-level action, it's undoing the cell addition. If you "redo" it puts the cell back. If you delete a cell like with cmd+backspace, then undo, it undoes the delete.

Your problem 2 is annoying, yes, we might be able to do a better job with the focus state of cells that are scrolled out of view (and thus not actually rendered and focused).

So for 1, I believe it is consistent, maybe just not obvious since the cell "add" may have been a long time ago, so you don't realize you're undoing it. And 3, it can be restored by "redo" which makes sense, this is reversing the action you just did.

Also, the local history does seem very inconsistent.

**Comment by repetitioestmaterstudiorum** [comment]
Thank you for the hint about redoing, I wasn't aware! That indeed solves (3).

**Comment by rebornix** [comment]
Let's use https://github.com/microsoft/vscode/issues/163943 to track problem 1&2 as both are talking about the same that it's not clear that users are undoing an operation they might not be aware of.
